scripts/casefold_db_dry_run.py

```diff
@@ -0,0 +1,199 @@
+
+    :return: None
+    """
+
+    # get every row where the local server is origin server and medium is email
+    origin_server = sydent.server_name
+    medium = "email"
+
+    cur = db.cursor()
+    cur.execute("BEGIN TRANSACTION;")
+    res = cur.execute(
+        "SELECT address, mxid, sgAssoc FROM global_threepid_associations WHERE medium = ?"
+        "AND originServer = ? ORDER BY ts DESC",
+        (medium, origin_server),
+    )
+
+    # dict that stores email address with mxid, email address, lookup hash, and
+    # signed association
+    associations: Dict[str, List[Tuple[str, str, str, str]]] = {}
+
+    # iterate through selected associations, casefold email, rehash it, re-sign the
+    # associations and add to associations dict
+    for address, mxid, sg_assoc in res.fetchall():
+        casefold_address = address.casefold()
+
+        # rehash the email since hash functions are case-sensitive
+        lookup_hash = calculate_lookup_hash(sydent, casefold_address)
+
+        # update signed associations with new casefolded address and re-sign
+        sg_assoc = json_decoder.decode(sg_assoc)
+        sg_assoc["address"] = address.casefold()
+        sg_assoc = json.dumps(
+            signedjson.sign.sign_json(
+                sg_assoc, sydent.server_name, sydent.keyring.ed25519
+            )
+        )
+
+        if casefold_address in associations:
+            associations[casefold_address].append(
+                (address, mxid, lookup_hash, sg_assoc)
+            )
+        else:
+            associations[casefold_address] = [(address, mxid, lookup_hash, sg_assoc)]
+
+    # list of arguments to update db with
+    db_update_args: List[Tuple[Any, str, str, str, str]] = []
+
+    # list of mxids to delete
+    to_delete: List[Tuple[str]] = []
+
+    # list of mxids and addresses to send emails to letting them know the mxid
+    # has been deleted
+    mxids: List[Tuple[Any, Any]] = []
+
+    for casefold_address, assoc_tuples in associations.items():
+        db_update_args.append(
+            (
+                casefold_address,
+                assoc_tuples[0][2],
+                assoc_tuples[0][3],
+                assoc_tuples[0][0],
+                assoc_tuples[0][1],
+            )
+        )
+
+        if len(assoc_tuples) > 1:
+            # Iterate over all associations except for the first one, since we've already
+            # processed it.
+            for address, mxid, _, _ in assoc_tuples[1:]:
+                to_delete.append((address,))
+                mxids.append((mxid, address))
+
+    if len(to_delete) > 0:
+        cur.executemany(
+            "DELETE FROM global_threepid_associations WHERE address = ?", to_delete
+        )
+
+    if len(db_update_args) > 0:
+        cur.executemany(
+            "UPDATE global_threepid_associations SET address = ?, lookup_hash = ?, sgAssoc = ? WHERE address = ? AND mxid = ?",
+            db_update_args,
+        )
+
+    cur.execute("ROLLBACK TRANSACTION;")
```

Comment:
though when we do come to actually dry-run, instead of rolling back, just don't execute the commands in the first place
(I probably misled you by saying we would like to know how long it'd take, sorry about that)

What would be useful though: print out the number of deletions and updates you'd have to do. That's usually a good enough indicator for how long a real run would do.
(`54 rows to delete, 27 rows to update` is something you can print regardless of dry vs real run)